mypy/test/teststubgen.py

```diff
@@ -804,6 +809,52 @@ def __init__(self, arg0: str) -> None:
             'def __init__(*args, **kwargs) -> Any: ...'])
         assert_equal(set(imports), {'from typing import overload'})
 
+    def test_cython(self) -> None:
+        pyx_source = """
+#cython: binding=True
+
+import typing
+
+def f(path: str, a: int = 0, b: bool = True) -> typing.List[str]:
```

Comment:
I have some questions about how Cython supports PEP 484 types:

* What if some argument has no type annotation?
* Do you know how Cython supports optional types? 
  * What would happen if this function would return None (with a list return type)? 
  * What if `f` is called with a `None` argument for `path`, for example?

(If Cython doesn't properly support optional types, we may not be able to use the types in stubs, at least not without some tweaks.)

Reply:
> What if some argument has no type annotation?

Then it's an `object`.

> What would happen if this function would return None (with a list return type)?

That's ok currently. Not sure if it's going to stay that way forever, given that optional types exist, but it's at least unlikely to change all too soon.

> What if f is called with a None argument for path, for example?

That's currently ok, too. `None` is generally allowed for builtin and extension types, unless explicitly excluded. That's probably also going to change at some point.

> If Cython doesn't properly support optional types

Cython currently only uses a subset of what PEP-484 allows, specifically: no container types, no optional types. Nothing that requires a `[]` or the `typing` module, basically. That's a missing feature, though, and will change at some point.

Maybe Cython should just implement `Optional[]` and apply `not None` semantics to types that use PEP-484 annotations. Probably just a little change in the type parser (if it wasn't for the `typing` import tracking, but we could ignore that, I guess).

Reply:
If Cython doesn't yet ensure that optional types are used consistently, we can't easily know whether an argument should allow `None` values in a stub, and whether a function might return `None`. A conservative approximation would be to *always* use an optional type for an argument, and use `Union[<type>, Any]` for the return type. The latter is the best way we have of saying that a function returns `<type>` and potentially `None` (but we don't know for sure). These are still better than using plain `Any` or `object` as the types. 

Also, we don't usually want to use `object` as a return type, and it will usually result in many false positives. Instead, it should probably be replaced with `Any`.

> Maybe Cython should just implement Optional[] and apply not None semantics to types that use PEP-484 annotations.

This would be great, since optional types tend to be used a lot in PEP 484 annotated code. Mypy also didn't properly support optional types in the early days (and we still allow disabling strict None checks, but this is mostly for legacy code).

Reply:
> Also, we don't usually want to use object as a return type, and it will usually result in many false positives. Instead, it should probably be replaced with Any.

For regular functions I don't think Cython enforces anything about the return type at all. Therefore, from a Cython point of view it should probably always be treated as `Any` (although I guess you might chose to trust the user and use their annotation?).

For `cdef` functions is does have specified return types, but those are only visible from C anyway so probably not relevant to mypy.

Reply:
> For regular functions …

You probably mean plain Python `def` functions here. For `cpdef` / `ccall` functions (and `cdef` / `cfunc`, obviously), i.e. anything that has a C visible return type, the return types are enforced and checked like for an assignment.

I agree that mypy should believe the user annotations, with the above caveat regarding `None`.